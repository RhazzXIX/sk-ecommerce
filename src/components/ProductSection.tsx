"use client";
import { CSSProperties } from "react";
import { useState } from "react";
import { FixedSizeList as List } from "react-window";
import InfiniteLoader from "react-window-infinite-loader";
import ProductCard from "@/components/ProductCard";

export default function ProductSection(props) {
  const [data, setData] = useState<product[]>([]);
  const [itemCount, setItemCount] = useState(100);
  let loading: Boolean;

  const Row = ({ index, style }: { index: number; style: CSSProperties }) => {
    const item: product = data[index];
    let content;
    if (!item) return null;
    return <ProductCard key={item.id} style={style} product={item} />;
  };

  const isItemLoaded = (index: number) => {
    return index <= data.length - 1;
  };

  const loadMoreItems = async (
    startIndex: number,
    stopIndex: number
  ): Promise<void> => {
    return new Promise((resolve) => {
      if (loading) return;
      loading = true;
      fetch(`https://dummyjson.com/products?skip=${data.length}&limit=${15}`)
        .then((res) => res.json())
        .then((newItems) => {
          if (data.length === 100) return;
          setData((prevData) => [...prevData, ...newItems.products]);

          resolve();
          loading = false;
        })
        .catch((error) => {
          console.error(error);
        });
    });
  };
  return (
    <section>
      <header></header>
      <InfiniteLoader
        isItemLoaded={isItemLoaded}
        itemCount={itemCount}
        loadMoreItems={loadMoreItems}
      >
        {({ onItemsRendered, ref }) => (
          <List
            itemCount={itemCount}
            onItemsRendered={onItemsRendered}
            ref={ref}
            height={800}
            itemSize={200}
            width={"100%"}
          >
            {Row}
          </List>
        )}
      </InfiniteLoader>
    </section>
  );
}
