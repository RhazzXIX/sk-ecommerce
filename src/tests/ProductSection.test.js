import { render, screen, waitFor } from "@testing-library/react";
import ProductSection from "../components/ProductSection";
import "@testing-library/jest-dom";
import { act } from "react-dom/test-utils";

global.fetch = jest.fn();

const newItems = {
  products: [
    {
      brand: "Apple",
      category: "smartphones",
      description: "An apple mobile which is nothing like apple",
      discountPercentage: 12.96,
      id: 1,
      price: 549,
      rating: 4.69,
      stock: 94,
      title: "iPhone 9",
      images: [
        "https://i.dummyjson.com/data/products/11/1.jpg",
        "https://i.dummyjson.com/data/products/11/2.jpg",
        "https://i.dummyjson.com/data/products/11/3.jpg",
        "https://i.dummyjson.com/data/products/11/thumbnail.jpg",
      ],
    },
    {
      id: 100,
      title: "Crystal chandelier maria theresa for 12 light",
      description: "Crystal chandelier maria theresa for 12 light",
      price: 47,
      discountPercentage: 16,
      rating: 4.74,
      stock: 133,
      brand: "YIOSI",
      category: "lighting",
      images: [
        "https://i.dummyjson.com/data/products/11/1.jpg",
        "https://i.dummyjson.com/data/products/11/2.jpg",
        "https://i.dummyjson.com/data/products/11/3.jpg",
        "https://i.dummyjson.com/data/products/11/thumbnail.jpg",
      ],
    },
  ],
};

fetch.mockResolvedValue({
  json: () => newItems,
});

describe("ProductsSection component", () => {
  it("Renders on page", async () => {
    const { container } = render(<ProductSection />);
    await screen.findByText("iPhone 9");
    expect(container).toMatchSnapshot();
  });
});
