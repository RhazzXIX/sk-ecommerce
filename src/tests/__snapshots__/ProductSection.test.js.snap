// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsSection component Renders on page 1`] = `
<div>
  <section>
    <header />
    <div
      style="position: relative; height: 800px; width: 100%; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 20000px; width: 100%;"
      >
        <div
          class="card"
          style="position: absolute; left: 0px; top: 0px; height: 200px; width: 100%;"
        >
          <img
            alt=""
            data-nimg="1"
            decoding="async"
            height="0"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F11%2F1.jpg&w=16&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F11%2F1.jpg&w=16&q=75 1x"
            style="color: transparent;"
            width="0"
          />
          <h4>
            iPhone 9
          </h4>
          <p>
            An apple mobile which is nothing like apple
          </p>
          <p>
            $ 
            549
            .00
          </p>
          <div
            class="qtyContainer"
          >
            <label
              for="qty"
            >
              Qty:
              <input
                id="qty"
                inputmode="numeric"
                placeholder="0"
                type="text"
                value=""
              />
            </label>
            <button
              class="increment"
            >
              +
            </button>
            <button
              class="decrement"
            >
              -
            </button>
          </div>
          <p
            class="totalCost"
          >
            $ 
            0
            .00
          </p>
          <button
            class="addToCart"
          >
            Add to Cart
          </button>
        </div>
        <div
          class="card"
          style="position: absolute; left: 0px; top: 200px; height: 200px; width: 100%;"
        >
          <img
            alt=""
            data-nimg="1"
            decoding="async"
            height="0"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F11%2F1.jpg&w=16&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F11%2F1.jpg&w=16&q=75 1x"
            style="color: transparent;"
            width="0"
          />
          <h4>
            Crystal chandelier maria theresa for 12 light
          </h4>
          <p>
            Crystal chandelier maria theresa for 12 light
          </p>
          <p>
            $ 
            47
            .00
          </p>
          <div
            class="qtyContainer"
          >
            <label
              for="qty"
            >
              Qty:
              <input
                id="qty"
                inputmode="numeric"
                placeholder="0"
                type="text"
                value=""
              />
            </label>
            <button
              class="increment"
            >
              +
            </button>
            <button
              class="decrement"
            >
              -
            </button>
          </div>
          <p
            class="totalCost"
          >
            $ 
            0
            .00
          </p>
          <button
            class="addToCart"
          >
            Add to Cart
          </button>
        </div>
      </div>
    </div>
  </section>
</div>
`;
